### ISSUE TITLE
Permit vaulted passwords in extra variables through GUI and API.

### ISSUE THREAD

**Issue by dglinder** [ansible/awx#1992]
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

 - API
 - UI

##### SUMMARY

We have a workflow that passes some credentials (username and password) from Jenkins into Ansible Tower (passing "extra_var" via the API) to run a command on the remote systems.  These variables show up in the "Extra Variables" section and work fine when they are NOT encrypted with Vault.

When we use Vault to encrypt the variables and provide the appropriate Vault credentials to Tower, the process fails as the vaulted variables are not encrypted.

Attempting to simulate this in a simple example by pasting in vault encrypted variables results in the Tower GUI reporting "Error saving job template.  Error: Parse error, failed to parse variables."

See this RedHat support ticket for details:
    https://access.redhat.com/support/cases/#/case/02119398

Also see the attached screen shots showing the in-GUI extra variables setup attempt and error message.

##### ENVIRONMENT
* Tower version: 3.2.2
* Ansible version:  2.5.2
* Operating System: RHEL 7
* Web Browser: Firefox and Chrome (both currently latest released versions)

##### STEPS TO REPRODUCE

1. Create a vaulted variable:
    ansible-vault encrypt_string --name 'my_password' 'MySecretPa$$word!'
1. Ensure vault credential is setup with the proper password and added to the template.
1. Copy resulting "my_password" lines into the "Extra Variables" GUI, or pass it into the extra_vars section of the API call.

##### EXPECTED RESULTS

 * GUI - "Extra Variables" section populated and used during execution.
 * API - The extra variables section is used and variables are properly decoded during execution

##### ACTUAL RESULTS

 * GUI - The GUI reports "Error saving job template.  Error: Parse error, failed to parse variables." and will not save new Extra Variables block.
 * API - The variables are passed into API but are not decoded or used by the job.

##### ADDITIONAL INFORMATION

Screen shot showing vault password (screen shot "01"):

![01_-_command_line_ansible-vault_output](https://user-images.githubusercontent.com/1124371/41539409-5813f136-72d3-11e8-9dd0-bb803225b35f.png)

Screen shot showing Tower GUI error message (screen shot "02").

![02_-_job_template_with_extra_variables_and_credentials_and_error](https://user-images.githubusercontent.com/1124371/41539410-583687a0-72d3-11e8-8d7e-fadcf58cde0e.png)

**Comment by AlanCoding** [comment]
This is possible now, you have to use the JSON-compatible tags. Some instructions can be found in:

https://github.com/ansible/awx/issues/223

**Comment by dglinder** [comment]
@AlanCoding I don't believe this is resolved by using JSON tags for the extra variables section.  Following your steps from that page (May 11), but populating the Extra Variables section of the template doesn't work.

The JSON looks correct:
![image](https://user-images.githubusercontent.com/1124371/41604388-b8c582ac-73a4-11e8-8428-69d6b9b24f89.png)

Also in YAML format:
![image](https://user-images.githubusercontent.com/1124371/41604654-5eab436e-73a5-11e8-87c0-88726e0fd886.png)

As a test I have this playbook executed via Tower:

```
---
- hosts: all
  tasks:
    - name: Show what the_secret is...
      debug:
        msg: "the_secret is {{ the_secret }}"
        
    - name: Show what the_secret.__ansible_vault is...
      debug:
        msg: "the_secret is {{ the_secret.__ansible_vault }}"
```

But the `the_secret` variable shown in the debug is the vaulted 'blob', not the actual text:
![image](https://user-images.githubusercontent.com/1124371/41604574-31d62390-73a5-11e8-8d25-98e3ce6b3ab2.png)

**Comment by AlanCoding** [comment]
What version of Ansible do you have? That's the behavior I would expect with Ansible before that change landed.

**Comment by dglinder** [comment]
Ansible 2.5.3 (per the Tower 3.2.2 "About Tower" page).
![image](https://user-images.githubusercontent.com/1124371/41604902-eb11f758-73a5-11e8-84e6-3f1c840bab28.png)

**Comment by AlanCoding** [comment]
yeah, that version of Ansible core doesn't have this change. They're working on getting 2.6 out the door.

**Comment by dglinder** [comment]
Oh, ok thanks - my bad!

**Comment by odg0318** [comment]
@dglinder Did you resolve this issue? If resolved, what version ansible are you using?

@AlanCoding  I am using AWX 2.1.2.0 and Ansible 2.7.4. I tried a way in comment, https://github.com/ansible/awx/issues/1992#issuecomment-398424839, but an undecrypted variable was still shown on debugging.
```
TASK [Print foo] ***************************************************************
ok: [10.2.0.29] => {
    "msg": {
        "__ansible_vault": "$ANSIBLE_VAULT;1.1;AES256\n36373364616138303565303964643933643666636163353362666435656437396661623665656162\n6537313732316562646264393163383763346463303133620a623161336634623165666361333564\n63643565653537376230636635626464666334333731333162656562306362646438623361306232\n6338386564616532360a666163616433376537653839613131613833653231633961653866323739\n3261\n"
    }
}
```

I also tried to launch job template via the following API, POST `/api/v2/job_templates/[ID]/launch/`. The same result was shown.
```
curl -d '{"extra_vars": {"foo": {"__ansible_vault":"$ANSIBLE_VAULT;1.1;AES256\n36373364616138303565303964643933643666636163353362666435656437396661623665656162\n6537313732316562646264393163383763346463303133620a623161336634623165666361333564\n63643565653537376230636635626464666334333731333162656562306362646438623361306232\n6338386564616532360a666163616433376537653839613131613833653231633961653866323739\n3261\n"}}}' -XPOST -H "Content-Type: application/json"  -H "Authorization: Bearer XXX" http://awx_host/api/v2/job_templates/5/launch/
```

Please tell me how to use vault content in extra_vars in AWX.

**Comment by AlanCoding** [comment]
For use in JT extra_vars mentioned in the last comment, there is a proposed patch at https://github.com/ansible/awx/pull/2297

**Comment by odg0318** [comment]
As I understand, there is currently no solution, right?

**Comment by AlanCoding** [comment]
correct. They work if put in inventory variables (host or group as well), but not JT extra_vars.

**Comment by kladiv** [comment]
+1

**Comment by heckrock** [comment]
I need this as well

**Comment by obrienrobert** [comment]
It'd be great if this was possible.

**Comment by cherusk** [comment]
Is this viable by now or not? Most of the referred to PRs have have been ceased with a benign outcome, that's why I'm wondering ...

**Comment by AlexanderStohr** [comment]
I am using "Tower 3.4.2" with "Ansible 2.7.9" - and i am having those issue.

using this YAML prefix fails:
```
my_password: !vault | $ANSIBLE_VAULT;1.1;AES256
```
using this YAML prefix works:
```
my_password: >-
  !vault | $ANSIBLE_VAULT;1.1;AES256
```
switching the encoding in the working version towards JSON gives this prefix (this i assumed to be working as well):
```
{
 "my_password": "!vault | $ANSIBLE_VAULT;1.1;AES256 
```

As of now the Tower interpreter seems to insist purely on line-folding (see [YAML 1.2 reference from 2009](https://yaml.org/spec/1.2/spec.htm) ) for accepting the syntax - nothing else. (Was the first style ever a valid yaml syntax?)

**Comment by AlexanderStohr** [comment]
> using this YAML prefix works:
> 
> ```
> my_password: >-
>   !vault | $ANSIBLE_VAULT;1.1;AES256
> ```

clarification:
it works in the Ansible Tower web-GUI as an input.

when using such a sequence in a shell script block with a core command like this:
`echo "my_password={{ my_password }}"`
the result looks like this:
`my_password=!vault | ;1.1;AES256`
Evidently the $ANSIBLE_VAULT part from the raw string gets interpreted as an empty bash shell variable. No decoding of the vault data took place.

**Comment by AlexanderStohr** [comment]
maybe this one is explaining a bit:
https://github.com/ansible/ansible/issues/45922

"So, my issue here is that Ansible does not attempt to decrypt vars_secret_funky, and just returns it as-is.

This is the root cause of behavior reported in ansible/awx#223 (comment). While this issue could be addressed on the AWX side, we would prefer not to. The solution would be more simple if done here."

to my best understanding: the functionality is feasible but for supposed good design reasons developers don't want to provide it using the described way.

**Comment by AlexanderStohr** [comment]
There is a somewhat different explanation elsewhere: https://github.com/ansible/awx/pull/2297

"We don't support the !vault constructor in extra_vars. That would be fine, but we also pass extra_vars to Ansible as YAML. This is so that we can mark things as !unsafe. It was this sin we committed that left users with no option to put vaulted content in extra_vars. If we're doing such highly custom YAML conversations pre-run, then we should also detect Ansible's new special JSON syntax and apply the YAML constructor for this case."

this sounds to me like some design decision has closed the door to 1:1 vice versa conversions between YAML and JSON. re-opening would force to drop some other existing "feature" with all impact that might have it on folks using it already.

**Comment by AlexanderStohr** [comment]
i have now ticked for this work-around (=works for me):

created my needed bunch of vault-encrypted credentials in group_vars/all

**benefit**: one place to manage them all; easily stored in a repository (critical data wont vanish that easily); definitely works as it should (compared to EXTRA VARIABLES location that mainly wastes your time!)
**downside:** not per-job configuration of credentials method (i dont think i was in need for that in my setup)

**Comment by shanemcd** [comment]
Reopening this as the PR mentioned in https://github.com/ansible/awx/issues/1992#issuecomment-450853484 never got merged.

**Comment by TheRealHaoLiu** [comment]
related to https://github.com/ansible/awx/issues/1954
